### PR TITLE
166283052 function apps refactoring

### PR DESCRIPTION
**Pull request by rchessa**

### CODE REVIEW

File: azurerm_function_app/main.tf [hunk 1]
```diff
@@ -1,11 +1,17 @@
 # FUNCTIONS
 
+data "azurerm_storage_account" "azurerm_functionapp_storage_account" {
+  name                = "${var.azurerm_functionapp_storage_account_name}"
+  resource_group_name = "${var.resource_group_name}"
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: I'd prefer to load the rg as data. Other parameters may be useful in other resources. Also, the rg name should come from the locals (copy from another variable/main file of other modules)

File: azurerm_function_app/main.tf [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+
 resource "azurerm_function_app" "azurerm_function_app" {
   name                      = "${local.azurerm_functionapp_name}"
   location                  = "${var.location}"
   resource_group_name       = "${var.resource_group_name}"
   app_service_plan_id       = "${azurerm_app_service_plan.azurerm_app_service_plan.id}"
-  storage_connection_string = "${azurerm_storage_account.azurerm_functionapp_storage_account.primary_connection_string}"
+  storage_connection_string = "${data.azurerm_storage_account.azurerm_functionapp_storage_account.primary_connection_string}"
   client_affinity_enabled   = false
   version                   = "~1"
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: shouldn't be 2?

File: azurerm_function_app/main.tf [hunk 3]
```diff
@@ -101,27 +107,27 @@ resource "azurerm_app_service_plan" "azurerm_app_service_plan" {
   }
 }
 
-# STORAGE ACCOUNT
-resource "azurerm_storage_account" "azurerm_functionapp_storage_account" {
-  name                = "${local.azurerm_functionapp_storage_account_name}"
-  resource_group_name = "${var.resource_group_name}"
-  location            = "${var.location}"
+# # STORAGE ACCOUNT
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: No upper case

File: azurerm_function_app/main.tf [hunk 4]
```diff
@@ -101,27 +107,27 @@ resource "azurerm_app_service_plan" "azurerm_app_service_plan" {
...
+# resource "azurerm_storage_account" "azurerm_functionapp_storage_account" {
+#   name                = "${local.azurerm_functionapp_storage_account_name}"
+#   resource_group_name = "${var.resource_group_name}"
+#   location            = "${var.location}"
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: Comments are for comments only :) No code commented in the repo besides exceptions please

File: azurerm_function_app/variables.tf [hunk 5]
```diff
@@ -15,15 +15,6 @@ variable "resource_name_prefix" {
...
-  # Define resource names based on the following convention:  # {azurerm_resource_name_prefix}-RESOURCE_TYPE-{environment}
-
-  azurerm_functionapp_name = "${var.resource_name_prefix}-functions-${var.environment}"
-
-  azurerm_functionapp_storage_account_name = "${var.resource_name_prefix}funcstorage${var.environment}"
-  azurerm_app_service_plan_name            = "${var.resource_name_prefix}-app-${var.environment}"
-}
-
 # TF_VAR_ADB2C_TENANT_ID
 variable "ADB2C_TENANT_ID" {
```
Comment (LucaPrete) [inline azurerm_function_app/variables.tf]: is there a reason for this to be upper case?

File: azurerm_function_app/main.tf [hunk 6]
```diff
@@ -1,11 +1,17 @@
 # FUNCTIONS
 
+data "azurerm_storage_account" "azurerm_functionapp_storage_account" {
+  name                = "${var.azurerm_functionapp_storage_account_name}"
+  resource_group_name = "${var.resource_group_name}"
+}
+
+
 resource "azurerm_function_app" "azurerm_function_app" {
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: please add at the end of each resource block the environment tag

File: azurerm_function_app/variables.tf [hunk 7]
```diff
@@ -15,15 +15,6 @@ variable "resource_name_prefix" {
...
-locals {
-  # Define resource names based on the following convention:  # {azurerm_resource_name_prefix}-RESOURCE_TYPE-{environment}
-
-  azurerm_functionapp_name = "${var.resource_name_prefix}-functions-${var.environment}"
-
-  azurerm_functionapp_storage_account_name = "${var.resource_name_prefix}funcstorage${var.environment}"
-  azurerm_app_service_plan_name            = "${var.resource_name_prefix}-app-${var.environment}"
-}
-
 # TF_VAR_ADB2C_TENANT_ID
```
Comment (LucaPrete) [inline azurerm_function_app/variables.tf]: What's the need for this comment?

File: azurerm_function_app/variables.tf [hunk 8]
```diff
@@ -129,3 +120,14 @@ variable "account_replication_type" {
...
+
+variable "azurerm_functionapp_storage_account_name" {
+  description = "The storage account name used by function_app"
+}
+
+locals {
+  # Define resource names based on the following convention:  # {azurerm_resource_name_prefix}-RESOURCE_TYPE-{environment}
+  azurerm_functionapp_name = "${var.resource_name_prefix}-functions-${var.environment}"
+  # = "${var.resource_name_prefix}funcstorage${var.environment}"
+  azurerm_app_service_plan_name            = "${var.resource_name_prefix}-app-${var.environment}"
```
Comment (LucaPrete) [inline azurerm_function_app/variables.tf]: equals not aligned

File: azurerm_function_app/main.tf (hunk 1 above)
Comment (rchessa) [reply azurerm_function_app/main.tf]: Ok I fix it.

File: azurerm_function_app/variables.tf (hunk 5 above)
Comment (rchessa) [reply azurerm_function_app/variables.tf]: It was inherited from the azure.tf file, but it's not a problem to change it

File: azurerm_function_app/variables.tf [hunk 9]
```diff
@@ -129,3 +120,15 @@ variable "account_replication_type" {
...
+  description = "The storage account name used by function_app"
+}
+
+locals {
+  # Define resource names based on the following convention:  # {azurerm_resource_name_prefix}-RESOURCE_TYPE-{environment}
+  azurerm_functionapp_name      = "${var.resource_name_prefix}-functions-${var.environment}"
+  azurerm_resource_group_name   = "${var.resource_name_prefix}-${var.environment}-rg"
+  azurerm_storage_account_name  = "${var.resource_name_prefix}${var.environment}sa${var.storage_account_name}"
+  azurerm_app_service_plan_name = "${var.resource_name_prefix}-app-${var.environment}"
+}
```
Comment (LucaPrete) [inline azurerm_function_app/variables.tf]: aggiungere riga vuota alla fine del file se non presente

File: azurerm_function_app/main.tf [hunk 10]
```diff
@@ -96,32 +102,7 @@ resource "azurerm_app_service_plan" "azurerm_app_service_plan" {
   sku {
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: @rchessa please make these variables

File: azurerm_function_app/variables.tf [hunk 11]
```diff
@@ -15,17 +15,8 @@ variable "resource_name_prefix" {
   description = "The prefix used to name all resources created."
```
Comment (LucaPrete) [inline azurerm_function_app/variables.tf]: There are few things that need to be fixed that are not directly visible from this page:
* Remove all commented/unused code
* I'd like to discuss the usage of the variable function. Was it used for debug? Can we remove it?

File: azurerm_function_app/main.tf [hunk 12]
```diff
@@ -96,32 +102,7 @@ resource "azurerm_app_service_plan" "azurerm_app_service_plan" {
...
     tier = "Standard"
 
-    # see https://azure.microsoft.com/en-en/pricing/details/app-service/
+    # See https://azure.microsoft.com/en-en/pricing/details/app-service/
     size = "S1"
```
Comment (LucaPrete) [inline azurerm_function_app/main.tf]: These should be variables

**Pull request merged**